src/coreclr/vm/codeman.cpp

```diff
@@ -5912,6 +5912,86 @@ int NativeUnwindInfoLookupTable::LookupUnwindInfoForMethod(DWORD RelativePc,
     return -1;
 }
 
+int HotColdMappingLookupTable::LookupMappingForMethod(ReadyToRunInfo* pInfo, ULONG MethodIndex)
+{
+    CONTRACTL {
+        NOTHROW;
+        GC_NOTRIGGER;
+        SUPPORTS_DAC;
+    } CONTRACTL_END;
+
+    if (pInfo->m_nHotColdMap == 0)
+    {
+        return -1;
+    }
+
+    // Casting the lookup table's size to an int is safe:
+    // We index the RUNTIME_FUNCTION table with ints, and the lookup table
+    // contains a subset of the indices in the RUNTIME_FUNCTION table.
+    // Thus, |lookup table| <= |RUNTIME_FUNCTION table|.
+    _ASSERTE(pInfo->m_nHotColdMap <= pInfo->m_nRuntimeFunctions);
+    const int nLookupTable = (int)(pInfo->m_nHotColdMap);
+
+    // The lookup table contains pairs of hot/cold indices, and thus should have an even size.
+    _ASSERTE((nLookupTable % 2) == 0);
+    int high = ((nLookupTable - 1) / 2);
+    int low  = 0;
+
+    const int indexCorrection = (int)(MethodIndex < pInfo->m_pHotColdMap[0]);
+
+    // Binary search the lookup table.
+    // Use linear search once we get down to a small number of elements
+    // to avoid binary search overhead.
```

Comment:
Can you please elaborate why linear search is considered to be faster than binary search for low element counts? At the first look the bisection loop seems much simpler than the linear scan part. I know that bubble sort is faster than quick sort when the number of elements is small but I don't see how it translates to mere lookup. Removal of the linear loop would also likely make the code easier to reason about.

Reply:
I don't have insight into the performance difference, but it's worth noting we also use linear search for low elements counts when searching the runtime function table [here](https://github.com/dotnet/runtime/blob/6c949f5d65f1bb89297dd19e5a107dc1ec420507/src/coreclr/vm/codeman.cpp#L5878); I decided to mimic this implementation for consistency, with the expectation that this will change if we decide to redesign the hot/cold mapping data structure (see [runtimelab/1903](https://github.com/dotnet/runtimelab/issues/1903)).

Reply:
Thanks Aman for clarifying and apologies for not mentioning you in my overall review response, I originally didn't realize this PR includes your JIT changes, I was under the impression that these have already been merged in. I have fixed this in the PR review feedback.

Reply:
No worries -- thank you for the review!

> I originally didn't realize this PR includes your JIT changes, I was under the impression that these have already been merged in

I admit my workflow is a bit confusing: Almost all of my [JIT work](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/hot-cold-splitting.md) was done in dotnet/runtime (except for [some fixes](https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/jit/hot-cold-splitting.md) to unwind info for cold EH funclets), but I did make a few contributions to the runtime and Crossgen2 that made it into this PR. Splitting support is a bit disjoint right now between the two repos, but that will fix itself if we decide to merge this PR.

Reply:
Lookup is well known to be faster for small numbers of elements. It turns out that the CPU pipelining optimizations work MUCH better with linear scan than binary search.